app/models/concerns/digital_object_concerns/save_behavior/resource_imports.rb

```diff
@@ -5,24 +5,166 @@ module SaveBehavior
+      def process_resource_imports(lock_object = nil)
+        self.resource_import_attributes.each do |resource_import_name|
+          process_resource_import(lock_object, resource_import_name)
+        end
+        yield if block_given?
+        # If everything went well, then clear the file imports
+        clear_resource_imports
+      rescue StandardError => e
+        # The import failed, so we want to undo all file copy operations that occurred
+        undo_new_resource_file_copies
+        raise e # pass along the exception to trigger object rollback
+      end
+
+      # Processes the data for the resource_import at key resource_import_name (e.g. 'access') and
+      # upon success assigns a new resource at key resource_import_name.
+      def process_resource_import(lock_object, resource_import_name)
+        resource_import = self.resource_imports[resource_import_name]
+        return if resource_import.nil? # nothing to import for this resource
+
+        # Immediately assign this resource, with location, to resources hash, even though we
+        # still need to add a couple of other properties to it later in the method.
+        # This will enable copy undo to work properly in later code.
+        self.resources[resource_import_name] = new_resource_for_import(resource_import, resource_import_name)
+
+        # Next we're going to do a checksum comparison, if a checksum was given.
+        # We want to fail quickly if an invalid checksum was given, so we'll check the given
+        # value first.
+        provided_hexdigest = hexgidest_from_import_checksum(resource_import)
+        provided_file_size = resource_import.file_size.present? ? resource_import.file_size : nil
+
+        hexdigest, file_size = analyze_and_optionally_copy_resource_import(resource_import, self.resources[resource_import_name], lock_object)
+
+        # If checksum was provided, let's compare it to our calculated value
+        validate_provided_hexdigest!(provided_hexdigest, hexdigest)
+
+        # If file size was provided, let's compare it to our calculated file size
+        validate_provided_file_size!(provided_file_size, file_size)
+
+        # If we got here, everything worked!  Let's convert the checksum to our internally-stored
+        # URN format and assign it and the file_size to the newly-created resource object.
+        self.resources[resource_import_name].checksum = "urn:sha256:#{hexdigest}"
+        self.resources[resource_import_name].file_size = file_size
+      end
+
+      def new_resource_for_import(resource_import, resource_import_name)
+        original_file_path = original_file_path_from_import(resource_import)
+
+        Hyacinth::DigitalObject::Resource.new(
+          location: location_from_import(resource_import, resource_import_name),
+          original_file_path: original_file_path,
+          media_type: media_type_from_import(File.basename(original_file_path)),
+          is_new: true # used during rollback
```

Comment:
Just not added when the object is loaded -- and now that you mention it, that might be a problem if we call `process_resource_imports` twice for the same instance of an object (which wouldn't normally happen, but theoretically could in a rake task or via rails console).

I was going back and forth for a while about whether is_new should default to true or false for a Resource instance.  Defaulting to true is a little more intuitive ("it's new because it's a new instance"), but defaulting to false is a little safer when it comes to handing deletion of resources in the case of rollback (since only is_new=true resources are deleted on rollback).  Do you have a preference between the two defaults, true or false?

Either way, I'll make sure to mark things as is_new=false after a successful import.

Reply:
My instinct would be to default `is_new` to `false` unless it's specifically set to `true`. But it makes me think, maybe it should have a default? Maybe it should be specifically set?

Marking things as `is_new=false`, would happen after a successful save, right? After we make sure we don't have to rollback?

Reply:
Yeah, requiring a specifically set value is a good idea.  Seems safer.  I'll do that.  And that's correct about `is_new=false` after a successful save.

Reply:
Actually, after going through and updating a bunch of tests, I'm realizing that `is_new` with value of `false` applies in all cases other than the import process, so I think defaulting to false is the way to go after all, and not actually specifically setting it unless we're in a case where it should be true (i.e. just the import case), so I'm updating some code now to revert the local changes I started making.  Does that sound okay?

Reply:
Yea, that sounds good!

Reply:
Great, thanks!